File: Data Structures & Algorithms/three-integer-sum/submission-6.cpp
#include <algorithm>
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> solutions;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++) {
            int target = nums[i];

            if (target > 0) {
                break;
            }
            if (i > 0 && target == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = nums.size() - 1;

            while (left < right) {
                int sum = nums[left] + nums[right] + target;
                if (sum == 0) {
                    solutions.push_back({nums[left], nums[right], target});
                    left++;
                    right--;
                    while (nums[left] == nums[left - 1] && left < right) {
                        left++;
                    }
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return solutions;
    }
};

File: Data Structures & Algorithms/binary-search/submission-6.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)

        while left < right:
            mid = left + (right - left)// 2 #prevent overflow when l and r are large
            num = nums[mid]

            if num == target:
                return mid
            elif num > target:
                right = mid
            else:
                left = mid + 1
        
        return -1